gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch

```diff
@@ -0,0 +1,2122 @@
++  gint max_delayed_frames;
++  GstClockTime latency;
++
++  /* FIXME get a real value from the encoder, this is currently not exposed */
++  max_delayed_frames = 5;
++
++  if (info->fps_n) {
++    latency = gst_util_uint64_scale_ceil (GST_SECOND * info->fps_d,
++        max_delayed_frames, info->fps_n);
++  } else {
++    /* FIXME: Assume 25fps. This is better than reporting no latency at
++     * all and then later failing in live pipelines
++     */
++    latency = gst_util_uint64_scale_ceil (GST_SECOND * 1,
++        max_delayed_frames, 25);
++  }
++
++  GST_INFO_OBJECT (encoder,
++      "Updating latency to %" GST_TIME_FORMAT " (%d frames)",
++      GST_TIME_ARGS (latency), max_delayed_frames);
++
++  gst_video_encoder_set_latency (GST_VIDEO_ENCODER (encoder), latency, latency);
++}
++
++static gboolean
++gst_svthevc_enc_set_format (GstVideoEncoder * video_enc,
++    GstVideoCodecState * state)
++{
++  GstSvtHevcEnc *encoder = GST_SVTHEVC_ENC (video_enc);
++  GstVideoInfo *info = &state->info;
++  GstCaps *allowed_caps;
++
++  /* If the encoder is initialized, do not reinitialize it again if not
++   * necessary */
++  if (encoder->svt_handle) {
++    GstVideoInfo *old = &encoder->input_state->info;
++
++    if (info->finfo->format == old->finfo->format
++        && info->width == old->width && info->height == old->height
++        && info->fps_n == old->fps_n && info->fps_d == old->fps_d
++        && info->par_n == old->par_n && info->par_d == old->par_d) {
++      gst_video_codec_state_unref (encoder->input_state);
++      encoder->input_state = gst_video_codec_state_ref (state);
++      return TRUE;
++    }
++
++    /* clear out pending frames */
++    gst_svthevc_enc_flush_frames (encoder, TRUE);
++  }
++
++  if (encoder->input_state)
++    gst_video_codec_state_unref (encoder->input_state);
++  encoder->input_state = gst_video_codec_state_ref (state);
++
++  allowed_caps = gst_pad_get_allowed_caps (GST_VIDEO_ENCODER_SRC_PAD (encoder));
++  if (allowed_caps) {
++    GstStructure *s;
++    const gchar *profile;
++    const gchar *level;
++    const gchar *tier;
++
++    GST_LOG_OBJECT (encoder, "allowed caps %" GST_PTR_FORMAT, allowed_caps);
++
++    if (gst_caps_is_empty (allowed_caps)) {
++      gst_caps_unref (allowed_caps);
++      return FALSE;
++    }
++
++    allowed_caps = gst_caps_make_writable (allowed_caps);
++    allowed_caps = gst_caps_fixate (allowed_caps);
++    s = gst_caps_get_structure (allowed_caps, 0);
++
++    profile = gst_structure_get_string (s, "profile");
++    if (profile) {
++      if (g_str_has_prefix (profile, "main-10")) {
++        encoder->profile = 2;
++      } else if (g_str_has_prefix (profile, "main-4:4:4")) {
++        encoder->profile = 4;
++      } else if (g_str_has_prefix (profile, "main")) {
++        encoder->profile = 1;
++      } else {
++        g_assert_not_reached ();
++      }
++    }
++
++    level = gst_structure_get_string (s, "level");
```

Comment:
I don't think you will get level and tier here, since they are not part of src_factory.
They will always be NULL, even if you specify: video/x-h265,profile=main-4:4:4,level=3.1

Reply:
I am not sure, if there is no specify field in src_factory (e.g. level), it will be provided in the allowed caps because it does not intersect caps. I'll check this.